### PR TITLE
More tokenizer's APIs cleanup

### PR DESCRIPTION
**Pull request by tarekgh**
This change include the following:
- The decoder abstracted class has been removed. During the implementation of Tiktoken, it became evident that the decoder abstracted interface will not be suited for decoding operation in models like Tiktoken. This is because this specific tokenizer requires decoding the IDs collectively rather than decoding each ID separately. We have already introduced the Decode method in the Model abstraction, which has proven effective and will also benefit all tokenizers, as tokenizer models can now leverage their internal state to perform more sophisticated decoding. The changes made include adding decoding code for the Bpe model, resulting in significantly improved results compared to before.
- Addressing the handling of unknown tokens in the Bpe model to ensure consistency with the approach used in the Huggingface Bpe model.

### CODE REVIEW

**Comment by tarekgh** [comment]
CC @ericstj @michaelgsharp

File: src/Microsoft.ML.Tokenizers/Model/BPE.cs [hunk 1]
```diff
@@ -35,25 +37,20 @@ public sealed class Bpe : Model
...
+                    return;
                 }
 
-
-                if (value is not null)
+                if (!_vocab.TryGetValue(value, out int id))
                 {
-                    _vocab[new StringSpanOrdinalKey(value)] = 0;
-                    VocabReverse[0] = value;
+                    throw new InvalidOperationException($"Unknown Token '{value}' Out Of Vocabulary.");
```
Comment (ericstj) [inline src/Microsoft.ML.Tokenizers/Model/BPE.cs:49]: nit
```suggestion
                    throw new InvalidOperationException($"Unknown Token '{value}' was not present in '{nameof(Vocab)}'.");
```

**Review by ericstj — APPROVED** [review]
Looks good.  Just aa couple small suggestions.

**Review by michaelgsharp — APPROVED** [review]
LGTM.

**Pull request merged**